Source/Core/VideoCommon/UberShaderVertex.cpp

```diff
@@ -176,8 +176,45 @@ ShaderCode GenVertexShader(APIType api_type, const ShaderHostConfig& host_config
                "\n");
 
   // Hardware Lighting
-  WriteVertexLighting(out, api_type, "pos.xyz", "_norm0", "rawcolor0", "rawcolor1", "o.colors_0",
-                      "o.colors_1");
+  out.WriteFmt("// xfmem.numColorChans controls the number of color channels available to TEV,\n"
+               "// but we still need to generate all channels here, as it can be used in texgen.\n"
+               "// Cel-damage is an example of this.\n"
+               "float4 vertex_color_0, vertex_color_1;\n"
+               "\n");
+  out.WriteFmt("// To use color 1, the vertex descriptor must have color 0 and 1.\n"
+               "// If color 1 is present but not color 0, it is used for lighting channel 0.\n"
+               "bool use_color_1 = ((components & {0}) == {0}); // VB_HAS_COL0 | VB_HAS_COL1\n",
+               VB_HAS_COL0 | VB_HAS_COL1);
+
+  out.WriteFmt("for (uint color = 0; color < {}; color++) {\n", NUM_XF_COLOR_CHANNELS);
+  out.WriteFmt("  if ((color == 0 || use_color_1) && (components & ({} << color)) != 0) {\n",
+               VB_HAS_COL0);
+  out.WriteFmt(
+      "    float4 color_value;\n"
+      "    // Use color0 for channel 0, and color1 for channel 1 if both colors 0 and 1 are "
+      "present.\n"
+      "    if (color == 0u)\n"
+      "      vertex_color_0 = rawcolor0;\n"
+      "    else\n"
+      "      vertex_color_1 = rawcolor1;\n"
+      "  } else if (color == 0 && (components & %uu) != 0) {\n",
```

Comment:
braces need to be doubled up in fmt contexts, like `{{`, which several other added lines don't seem to be doing. A format specifier is also using printf-like formatting, which fmt doesn't use (this, too, seems to be an issue on multiple lines of added code).

Reply:
Thanks for mentioning this.  I was careful about it for lines where there were merge conflicts (redoing the changes off of the fmt version), but I must have missed some in code that was completely new.